lib/integrations/externalcloudaudit/error_counter.go

```diff
@@ -0,0 +1,403 @@
+	}
+	c.categories = []errorCategory{
+		{
+			alertName:    eventEmitFailureClusterAlert,
+			alertMessage: eventEmitFailureClusterAlertMsgTemplate,
+			// Raise a cluster alert if the recent error rate reaches 40%.
+			alertThreshold: 0.4,
+			// Don't alert until there have been at least 10 errors.
+			minErrorsForAlert: 10,
+			// Clear the alert if the recent error rate is below 5%.
+			clearAlertThreshold: 0.05,
+			// Don't clear the alert unless there have been at least 10
+			// successes.
+			minSuccessesForClear: 10,
+			count:                &c.emits,
+		},
+		{
+			alertName:            eventSearchFailureClusterAlert,
+			alertMessage:         eventSearchFailureClusterAlertMsgTemplate,
+			minErrorsForAlert:    4,
+			alertThreshold:       0.4,
+			minSuccessesForClear: 4,
+			clearAlertThreshold:  0.05,
+			count:                &c.searches,
+		},
+		{
+			alertName:            sessionUploadFailureClusterAlert,
+			alertMessage:         sessionUploadFailureClusterAlertMsgTemplate,
+			minErrorsForAlert:    4,
+			alertThreshold:       0.4,
+			minSuccessesForClear: 8,
+			clearAlertThreshold:  0.05,
+			count:                &c.uploads,
+		},
+		{
+			alertName:            sessionDownloadFailureClusterAlert,
+			alertMessage:         sessionDownloadFailureClusterAlertMsgTemplate,
+			minErrorsForAlert:    4,
+			alertThreshold:       0.4,
+			minSuccessesForClear: 2,
+			clearAlertThreshold:  0.05,
+			count:                &c.downloads,
+		},
+	}
+	return c
+}
+
+// WrapAuditLogger returns an [events.AuditLogger] that will forward all calls
+// to [wrapped] and observe all errors encountered.
+func (c *ErrorCounter) WrapAuditLogger(wrapped events.AuditLogger) events.AuditLogger {
+	return newErrorCountingLogger(wrapped, &c.emits, &c.searches)
+}
+
+// WrapAuditLogger returns an [events.MultipartHandler] that will forward all
+// calls to [wrapped] and observe all errors encountered.
+func (c *ErrorCounter) WrapSessionHandler(wrapped events.MultipartHandler) events.MultipartHandler {
+	return newErrorCountingSessionHandler(wrapped, &c.uploads, &c.downloads)
+}
+
+// ObserveEmitError can be called to observe relevant event emit errors not
+// captured by WrapAuditLogger. In particular this should be used by the Athena
+// consumer which batches event writes to S3.
+func (c *ErrorCounter) ObserveEmitError(err error) {
+	c.emits.observe(err)
+}
+
+type errorCategory struct {
+	alertName            string
+	alertMessage         string
+	alertThreshold       float32
+	clearAlertThreshold  float32
+	minErrorsForAlert    uint64
+	minSuccessesForClear uint64
+	count                *errorCount
+}
+
+func (c *ErrorCounter) run(exitContext context.Context) {
+	ticker := c.clock.NewTicker(syncInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-exitContext.Done():
+			return
+		case <-ticker.Chan():
+		}
+		if err := c.sync(exitContext); err != nil {
+			log.WithError(err).Info("ErrorCounter sync failed")
+		}
+	}
+}
```

Comment:
refactored this so the draft configurator never starts the loop and the Configurator.ErrorCounter is documented to be nil for a draft Configurator (the whole draft configurator is a bit of a hack to get the creds just for the connection test, changing the API of it would require synchronized changes with teleport.e)